### PR TITLE
Check whether `maxiter` is integer

### PR DESCRIPTION
**Pull request by Psirus**
Check whether `maxiter` is an integer in optimize.zeros. This closes issue #10236.

### CODE REVIEW

**Comment by peterbell10** [comment]
Thanks for contributing @Psirus.

As @ev-br suggested in the issue, it might be better to do
```python
maxiter = operator.index(maxiter)
```

This will still raise on float input but would also allow e.g. numpy scalar values like `np.int32(10)`.

**Comment by Psirus** [comment]
Sure can do. I didn't know about `operator.index` and found it less obvious what this is supposed to do compared to explicitly throwing an exception. But I'm more than happy to change it.

**Review by peterbell10 — APPROVED** [review]

File: scipy/optimize/tests/test_zeros.py [hunk 1]
```diff
@@ -437,6 +437,11 @@ def test_newton_does_not_modify_x0(self):
         newton(np.sin, x0, np.cos)
         assert_array_equal(x0, x0_copy)
 
+    def test_maxiter_int_check(self):
+        for method in [zeros.bisect, zeros.newton, zeros.ridder, zeros.brentq, zeros.brenth, zeros.toms748]:
```
Comment (andyfaff) [inline scipy/optimize/tests/test_zeros.py]: nit, does this line exceed 80 chars?

File: scipy/optimize/tests/test_zeros.py [hunk 2]
```diff
@@ -437,6 +437,11 @@ def test_newton_does_not_modify_x0(self):
...
+            with pytest.raises(TypeError, match="'float' object cannot be interpreted as an integer"):
```
Comment (andyfaff) [inline scipy/optimize/tests/test_zeros.py]: same here, would be nice to keep to less than 80 chars.

**Comment by andyfaff** [comment]
Looks good to me now. Thank you @Psirus

**Pull request merged**